util/unifyfs/src/unifyfs.c

```diff
@@ -258,12 +279,176 @@ int main(int argc, char** argv)
+			cli_args.transfer_exe);
+		return -ENOENT;
+            }
+	    fclose(trans_exe_fp);
+	    trans_exe_fp = NULL;
+	    lines_in_manifest_file =
+	      unifyfs_lines_in_manifest_file(stage_in_filename);
+	    // fprintf(stderr,"lines_in_manifest_file returned >%d< lines\n",
+	    //		    lines_in_manifest_file);
+            if (lines_in_manifest_file < 0) {
+	        fprintf(stderr,
+			"ERROR! Cannot read stage-in manifest file >%s<!\n",
+		      stage_in_filename);
+                return -ENOENT;
+            }
+	    // possibly nest this in a if(debug) check in the future
+	    fprintf(stdout, "Stage-in manifest file contains %d lines.\n",
+		    lines_in_manifest_file);
+	} // if (stage_in_filename != NULL) {
+	else {
+            if (cli_args.transfer_exe) {
+	        fprintf(stderr,
+			"WARNING! You specified transfer_exe but not stage_in!\n");
+            }
+	}
         return unifyfs_start_servers(&resource, &cli_args);
-    } else if (action == ACT_TERMINATE) {
+    } // if (action == ACT_START)
+    else if (action == ACT_TERMINATE) {
+
+	if (cli_args.stage_in != NULL) {
+	    fprintf(stderr,
+		    "ERROR! called unifyfs terminate with --stage-in!\n");
+	    usage(1);
+            return -EINVAL;
+	}
+
+	stage_out_filename = cli_args.stage_out;
+	if (stage_out_filename != NULL) {
+            if (cli_args.script) {
+	        fprintf(stderr,
+			"WARNING! You are using a script with --stage-out.\n");
+	        fprintf(stderr, "This is unlikely to work.\n");
+            }
+            if (cli_args.transfer_exe == NULL) {
+	        fprintf(stderr,
+		      "ERROR!  You must specify --transfer-exe with --stage-out!\n");
+	        return -EINVAL;
+            }
+            if (trans_exe_fp == NULL) {
+	        fprintf(stderr,
+			"ERROR!  The trans exe file you specified >%s<doesn't exist!\n",
+			cli_args.transfer_exe);
+		return -ENOENT;
+            }
+	    fclose(trans_exe_fp);
+	    trans_exe_fp = NULL;
+	    lines_in_manifest_file =
+	      unifyfs_lines_in_manifest_file(stage_out_filename);
+            if (lines_in_manifest_file < 0) {
+	        fprintf(stderr,
+			"ERROR! Cannot read stage-out manifest file >%s<!\n",
+		      stage_in_filename);
+                return -ENOENT;
+            }
+	    // possibly nest this in a if(debug) check in the future
+	    fprintf(stdout, "Stage-in manifest file contains %d lines.\n",
+		    lines_in_manifest_file);
+	} else {
+            if (cli_args.transfer_exe) {
+	        fprintf(stderr,
+			"WARNING! You specified transfer_exe but not stage_out!\n");
+            }
+	}
         return unifyfs_stop_servers(&resource, &cli_args);
     } else {
         fprintf(stderr, "INTERNAL ERROR: unhandled action %d\n", (int)action);
         return -1;
     }
+} // int main(...
+
+
+// validates that the named manifest file exists and is readable
+// and returns the number of lines in it.
+// negative return value for error.
+int unifyfs_lines_in_manifest_file(char* manifest_filename)
```

Comment:
I have no clue why this is necessary. Please explain the motivation.

Reply:
I think it's prudent to check the manifest file for existence and to check its structure as much as is reasonable early on in the flow of unifyfs so that if it's clearly pathological, the user will have some indication that something's wrong.  If, for instance, the transfer command that does the stage-in finds something bad in the manifest file, I don't know that it has as good error reporting path to the user.  

Personally I think the positive confirmation that the front end found the manifest file and that it contains X lines is a nice confirmation that the unifyfs command got the --stage-in (or --stage-out) flag, found everything it was expecting, and reports that everything is fine.  This is a single-line, relatively unobtrusive way of reporting all of that along with a quantitative diagnostic value to match against what was expected.  

If you think we should wrap this in a if(debug) statement unless it find something pathological, then I suppose we could do that.

Reply:
I think we should just check for file existence at this point. Calling out to 'wc -l' only tells you that we got a file with some number of lines in it, it doesn't validate the contents in any way. I could pass a file full of '\n' characters and it would be happy. The format checking/parsing should be left to the transfer utility. If it detects a problem, it will generate a message to stderr, which is the same as what the unifyfs utility does for errors.

Reply:
Ok.  I wanted to do as much validation as possible at the front end, and you asked for all the validation to be done in unifyfs.c, not farther down, so I did as much as I could without actually parsing the file.  

As a user, if I have, say, a script that generates a manifest file from a directory for stage-in, and I'm expecting it to have 10 files, then it would be useful to me to have an initial debugging output that tells me "found 0 lines" or "found 5000 lines".  But if you think that manifest file validation should only be done in the transfer application, then I'll take this out once we've tested the unifyfs command and the transfer-stage command together.  I won't be in the next commit but it will be in the commit when we put in the final PR.